### PR TITLE
Latency measurement

### PR DESCRIPTION
**Pull request by AlexKordic**
**What does this pull request do? Explain your changes. (required)**

Add test to measure latency.

**Specific updates (required)**

- Refactor cmd/livepeer/livepeer.go to take external arguments
- Add integration test to bring up `B`, `O` & `T` nodes in same process and push several segments, retrieve result and measure latency

**How did you test each of these updates (required)**

`cd cmd/livepeer`

`go test . -v -count 1 -tags nvidia -run TestIntegration_TranscodingLatency`

observe `latency 15574.807013 ms; segment-duration=15s` log output

**What is still left to be done?**

Fix failing tests to target streaming workflow.

Segment transcode rejected leading to test failure.

```
broadcast.go:884] manifestID=13f63dab-5984-4580-b50d-7ab6753dc7cb nonce=3404987179542143874 seqNo=2 clientIP=127.0.0.1 No sessions available for segment
    assertion_compare.go:313: 
                Error Trace:    latency_test.go:193
                                                        asm_amd64.s:1571
                Error:          "0" is not greater than "0"
                Test:           TestIntegration_TranscodingLatency
                Messages:       [transcoding refused]
> recv ended with EOF chunks=0
```

Need to figure out why longer segments are rejected.

**Does this pull request close any open issues?**

**Checklist:**

- [ ] Read the [contribution guide](./doc/contributing.md)
- [ ] `make` runs successfully
- [ ] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: cmd/livepeer/livepeer.go [hunk 1]
```diff
@@ -69,6 +69,108 @@ var (
...
+	String(name string, value string, usage string) *string
+	Int(name string, value int, usage string) *int
+	Int64(name string, value int64, usage string) *int64
+	Float64(name string, value float64, usage string) *float64
+	Duration(name string, value time.Duration, usage string) *time.Duration
+	Visit(fn func(*flag.Flag))
+	GetCfg() LivepeerConfig
+}
+
+type RealProgramArguments struct{}
```
Comment (AlexKordic) [inline cmd/livepeer/livepeer.go:87]: Wondering is this good approach, could i use `flag` package directly?

Comment (AlexKordic) [reply cmd/livepeer/livepeer.go:87]: Should we move most of the logic out of `main` package?

File: server/mediaserver.go [hunk 2]
```diff
@@ -713,13 +713,28 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
...
+	if accept == "multipart/mixed" {
+		contentType := "multipart/mixed; boundary=" + boundary
+		w.Header().Set("Content-Type", contentType)
+	}
+	w.WriteHeader(http.StatusOK)
+	if f, ok := w.(http.Flusher); ok {
+		f.Flush()
+	}
 
+	body, err := common.ReadAtMost(r.Body, common.MaxSegSize)
```
Comment (AlexKordic) [inline server/mediaserver.go]: In streaming workflow we won't use `body`, instead operate only on `chunk := make([]byte, 16384)`s

File: server/mediaserver.go [hunk 3]
```diff
@@ -713,13 +713,28 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
...
+	ctx := r.Context()
+	remoteAddr := getRemoteAddr(r)
+	clog.Infof(ctx, "[latency] Start push request at url=%s addr=%s", r.URL.String(), remoteAddr)
+
+	boundary := common.RandName()
+	accept := r.Header.Get("Accept")
+	if accept == "multipart/mixed" {
+		contentType := "multipart/mixed; boundary=" + boundary
+		w.Header().Set("Content-Type", contentType)
+	}
```
Comment (AlexKordic) [inline server/mediaserver.go]: In streaming workflow we should use single output container with multiple video tracks and rely on Mist to repack for players.

File: cmd/livepeer/latency_test.go [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func toSeconds(duration time.Duration) float64 {
+	return float64(duration) / 1000000000.0
+}
+func fromSeconds(seconds float64) time.Duration {
+	return time.Duration(seconds * 1000000000.0)
+}
```
Comment (AlexKordic) [inline cmd/livepeer/latency_test.go:28]: @thomshutt I use these helper functions, is there another way that better matches the style of the go code?

Comment (thomshutt) [reply cmd/livepeer/latency_test.go:28]: Nope, seems pretty go-ish!

File: cmd/livepeer/latency_test.go [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+		latency := sending.lerpToReceiveTimeline(t, recvFraction, recvTimestamp)
+		r.Checkpoints = append(r.Checkpoints, TranscodingLatencyCheckpoint{
+			Milliseconds: latency,
+			ByteCount:    chunk.ByteCount,
+		})
+	}
+	return
+}
+
+func min(a, b int) int {
```
Comment (thomshutt) [inline cmd/livepeer/latency_test.go:118]: You could use `math.MinInt` for this

Comment (AlexKordic) [reply cmd/livepeer/latency_test.go:118]: Also my initial thought, turns out [MinInt](https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/math/const.go;l=43) is minimal integer value constant :)

Comment (thomshutt) [reply cmd/livepeer/latency_test.go:118]: Haha my bad!

File: server/mediaserver.go (hunk 3 above)
Comment (AlexKordic) [reply server/mediaserver.go]: Correction: remove the audio tracks altogether. Mist keeps the original track and guarantees sync between old and new tracks.

**Comment by Quintendos** [comment]
Ask @AlexKordic if this should be closed or merged.

**Comment by AlexKordic** [comment]
We used this one to measure latency initially, then @leszko e2e tests obsoleted this code. If we still need this test i should adopt it to match method we now use for integration tests.

**Pull request closed without merging**